### PR TITLE
Transition to tf.while_loops for iterative attacks

### PR DESCRIPTION
**Pull request by aidangomez**

### CODE REVIEW

**Comment by aidangomez** [comment]
@npapernot This passes checks now

File: cleverhans/attacks.py [hunk 1]
```diff
@@ -417,6 +420,10 @@ def generate(self, x, **kwargs):
             eta = adv_x - x
             from cleverhans.utils_tf import clip_eta
             eta = clip_eta(eta, self.ord, self.eps)
+            return i+1, eta
```
Comment (goodfeli) [inline cleverhans/attacks.py]: should be "i + 1" for PEP8

File: cleverhans/attacks.py [hunk 2]
```diff
@@ -417,6 +420,10 @@ def generate(self, x, **kwargs):
...
+
+        _, eta = tf.while_loop(
+            cond, body, [tf.zeros([]), eta], back_prop=False)
```
Comment (goodfeli) [inline cleverhans/attacks.py]: Usually our Attack classes make expressions that are differentiable. If you look at the tutorial, you see that for adversarial training we support an option of either back-proping through the attack or not, based on whether we wrap the output of the attack in a stop_gradient op. I'm not very familiar with the `while_loop` implementation. Is there a performance reason to use back_prop=False ahead of time?

File: cleverhans/attacks.py [hunk 3]
```diff
@@ -547,35 +557,40 @@ def generate(self, x, **kwargs):
...
-                adv_x = tf.clip_by_value(adv_x, self.clip_min, self.clip_max)
+                ax = tf.clip_by_value(ax, self.clip_min, self.clip_max)
 
-            adv_x = tf.stop_gradient(adv_x)
+            ax = tf.stop_gradient(ax)
+
+            return i+1, ax, m
+
+        _, adv_x, _ = tf.while_loop(
+            cond, body, [tf.zeros([]), adv_x, momentum], back_prop=False)
```
Comment (goodfeli) [inline cleverhans/attacks.py]: Same issue about back_prop=False here

File: cleverhans/attacks.py [hunk 4]
```diff
@@ -547,35 +557,40 @@ def generate(self, x, **kwargs):
...
+            ax = x + utils_tf.clip_eta(ax - x, self.ord, self.eps)
 
             if self.clip_min is not None and self.clip_max is not None:
-                adv_x = tf.clip_by_value(adv_x, self.clip_min, self.clip_max)
+                ax = tf.clip_by_value(ax, self.clip_min, self.clip_max)
 
-            adv_x = tf.stop_gradient(adv_x)
+            ax = tf.stop_gradient(ax)
+
+            return i+1, ax, m
```
Comment (goodfeli) [inline cleverhans/attacks.py]: PEP8

File: cleverhans/attacks.py [hunk 5]
```diff
@@ -1390,8 +1405,15 @@ def attack(self, x, y):
...
             eta = tf.zeros_like(x)
 
-        for i in range(self.nb_iter):
-            eta = self.attack_single_step(x, eta, y)
+        def cond(i, _):
+            return tf.less(i, self.nb_iter)
+
+        def body(i, e):
+            new_eta = self.attack_single_step(x, e, y)
+            return i+1, new_eta
```
Comment (goodfeli) [inline cleverhans/attacks.py]: PEP8

File: cleverhans/attacks.py [hunk 6]
```diff
@@ -1534,8 +1556,15 @@ def generate(self, x, g, **kwargs):
...
         eta = clip_eta(eta, self.ord, self.eps)
 
-        for i in range(self.nb_iter):
-            eta = self.attack_single_step(x, eta, g_feat)
+        def cond(i, _):
+            return tf.less(i, self.nb_iter)
+
+        def body(i, e):
+            new_eta = self.attack_single_step(x, e, g_feat)
+            return i+1, new_eta
```
Comment (goodfeli) [inline cleverhans/attacks.py]: PEP8

**Comment by goodfeli** [comment]
I made some comments about PEP8 but then while investigating while the tests passed, I found out PEP8 has changed and the PR respects the new version of PEP8. I deleted the comments related to PEP8.

**Comment by goodfeli** [comment]
This LGTM once we resolve the back_prop=False issue

**Comment by aidangomez** [comment]
@goodfeli Alrighty this should be g2g

**Pull request merged**